prov/util/src/util_mr_cache.c

```diff
@@ -182,16 +177,49 @@ bool ofi_mr_cache_flush(struct ofi_mr_cache *cache, bool flush_lru)
 		       entry->info.iov.iov_base, entry->info.iov.iov_len);
 
 		util_mr_uncache_entry_storage(cache, entry);
-		pthread_mutex_unlock(&mm_lock);
-
-		util_mr_free_entry(cache, entry);
-		pthread_mutex_lock(&mm_lock);
+		dlist_insert_tail(&entry->list_entry, &entries_to_free);
 
 	} while (!dlist_empty(&cache->lru_list) &&
 		 ((cache->cached_cnt >= cache_params.max_cnt) ||
 		  (cache->cached_size >= cache_params.max_size)));
 	pthread_mutex_unlock(&mm_lock);
 
+	while(!dlist_empty(&entries_to_free)) {
+		dlist_pop_front(&entries_to_free, struct ofi_mr_entry,
+				entry, list_entry);
+		dlist_init(&entry->list_entry);
+		util_mr_free_entry(cache, entry);
+	}
+
+	return true;
+}
+
+bool ofi_mr_cache_remove_dead_regions(struct ofi_mr_cache *cache)
+{
+	struct dlist_entry	entries_to_free;
+	struct ofi_mr_entry	*entry;
+
+	pthread_mutex_lock(&mm_lock);
+	if (dlist_empty(&cache->dead_region_list)) {
+		pthread_mutex_unlock(&mm_lock);
+		return false;
+	}
+
+	/* steal the whole list so we don't have to keep acquiring */
+	/* the mutex just for list management                      */
+
+	dlist_transfer(&cache->dead_region_list, &entries_to_free);
+	pthread_mutex_unlock(&mm_lock);
+
+	while (!dlist_empty(&entries_to_free)) {
+		dlist_pop_front(&entries_to_free, struct ofi_mr_entry,
+				entry, list_entry);
+		dlist_init(&entry->list_entry);
```

Comment:
No need to initialize objects about to be freed.

Reply:
i didnt think so either, but that was already in the code.